File: swift/Sources/TokenTrackerCore/Scanners/ScannerSupport.swift
//
//  ScannerSupport.swift
//  TokenTrackerCore
//
//  移植自 tokentracker/scanners/_util.py：文件指纹、字节游标增量读、
//  JSONL 迭代、zstd 解压、ISO 时间解析、会话标题提取。
//

import Foundation

public func expandPath(_ path: String) -> String {
    NSString(string: path).expandingTildeInPath
}

/// Python os.path.realpath（解析符号链接，如 /tmp → /private/tmp）。
public func realPath(_ path: String) -> String {
    if let resolved = Darwin.realpath(path, nil) {
        defer { free(resolved) }
        return String(cString: resolved)
    }
    return URL(fileURLWithPath: path).resolvingSymlinksInPath().path
}

// ------------------------------------------------------------ 文件指纹 ----

/// Python stat_key：{"m": mtime_ns, "s": size, "i": ino, "d": dev}
public struct StatKey: Equatable {
    public var m: Int64  // st_mtime_ns
    public var s: Int64  // size
    public var i: Int64  // ino
    public var d: Int64  // dev

    public init?(path: String) {
        var st = stat()
        guard stat(path, &st) == 0 else { return nil }
        m = Int64(st.st_mtimespec.tv_sec) * 1_000_000_000 + Int64(st.st_mtimespec.tv_nsec)
        s = Int64(st.st_size)
        i = Int64(st.st_ino)
        d = Int64(st.st_dev)
    }

    public var asDict: [String: Int64] { ["m": m, "s": s, "i": i, "d": d] }

    public init?(dict: [String: Any]?) {
        guard let dict else { return nil }
        func val(_ k: String) -> Int64? {
            (dict[k] as? Int64) ?? (dict[k] as? NSNumber)?.int64Value
        }
        guard let m = val("m"), let s = val("s"), let i = val("i"), let d = val("d") else { return nil }
        self.m = m; self.s = s; self.i = i; self.d = d
    }
}

/// Python changed()：只比对指纹键（游标可能携带 "o" 等附加字段）。
public func fingerprintChanged(cursor: [String: Any], path: String) -> Bool {
    guard let stored = cursor[path] as? [String: Any],
          let storedKey = StatKey(dict: stored),
          let current = StatKey(path: path) else { return true }
    return storedKey.m != current.m || storedKey.s != current.s
        || storedKey.i != current.i || storedKey.d != current.d
}

// ------------------------------------------------------------ JSONL ----

/// Python json.loads 一行 → [String: Any]（非 dict 返回 nil）。
public func parseJSONLine(_ data: Data) -> [String: Any]? {
    (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
}

/// 按 \n 切行并保留行号语义（Python 文件迭代只认 \n；空行计入行号）。
private func splitJSONLines(_ data: Data) -> [String] {
    let text = String(decoding: data, as: UTF8.self) // errors="replace" 对齐
    return text.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
}

/// iter_jsonl：yield (行号从 1 起, dict)。解析失败的行跳过；行号仍计入。
public func iterJSONL(_ path: String) -> [(Int, [String: Any])] {
    guard let data = FileManager.default.contents(atPath: path) else { return [] }
    var out: [(Int, [String: Any])] = []
    for (index, line) in splitJSONLines(data).enumerated() {
        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let obj = parseJSONLine(Data(trimmed.utf8)) else { continue }
        out.append((index + 1, obj))
    }
    return out
}

/// 字节级行读取器：read_jsonl_delta 的底座。
/// 返回 (行起始字节偏移, 行原始字节含 \n)；文件读尽返回 nil。
final class ByteLineReader {
    private let handle: FileHandle
    private var buffer: [UInt8] = []
    private var bufferStart: Int64   // buffer[0] 的文件绝对偏移
    private var index = 0            // buffer 内消费位置
    private var eof = false

    init?(path: String, offset: Int64) {
        guard let handle = FileHandle(forReadingAtPath: path) else { return nil }
        self.handle = handle
        self.bufferStart = offset
        do { try handle.seek(toOffset: UInt64(offset)) } catch { return nil }
    }

    deinit { try? handle.close() }

    private func fill() {
        guard !eof else { return }
        if index > 0 {
            buffer.removeFirst(index)
            bufferStart += Int64(index)
            index = 0
        }
        let chunk = handle.readData(ofLength: 64 * 1024)
        if chunk.isEmpty { eof = true } else { buffer.append(contentsOf: chunk) }
    }

    /// 返回 (绝对偏移, 行字节, 是否以 \n 结尾)。
    func readLine() -> (offset: Int64, data: Data, terminated: Bool)? {
        while true {
            if index < buffer.count,
               let nl = buffer[index...].firstIndex(of: 0x0A) {
                let start = bufferStart + Int64(index)
                let data = Data(buffer[index...nl])
                index = nl + 1
                return (start, data, true)
            }
            if eof {
                if index < buffer.count {
                    let start = bufferStart + Int64(index)
                    let data = Data(buffer[index...])
                    index = buffer.count
                    return (start, data, false)
                }
                return nil
            }
            fill()
        }
    }
}

/// read_jsonl_delta：从字节偏移（必须是行边界）增量读，只解析新增完整行。
/// 返回 (items, newOffset)；items 为 (行起始偏移, dict)。
/// 偏移失效（截断/轮转/不在行边界）返回 ([], -1)。
public func readJSONLDelta(path: String, offset: Int64) -> ([(Int64, [String: Any])], Int64) {
    guard let attrs = try? FileManager.default.attributesOfItem(atPath: path),
          let size = (attrs[.size] as? NSNumber)?.int64Value else { return ([], -1) }
    if offset < 0 || offset > size { return ([], -1) }
    if offset > 0 {
        // 校验 offset-1 必须是 \n（行边界）
        guard let handle = FileHandle(forReadingAtPath: path) else { return ([], -1) }
        defer { try? handle.close() }
        do { try handle.seek(toOffset: UInt64(offset - 1)) } catch { return ([], -1) }
        guard handle.readData(ofLength: 1) == Data([0x0A]) else { return ([], -1) }
    }
    guard let lineReader = ByteLineReader(path: path, offset: offset) else { return ([], -1) }
    var items: [(Int64, [String: Any])] = []
    var newOffset = offset
    while let (lineOffset, data, terminated) = lineReader.readLine() {
        guard terminated else {
            newOffset = lineOffset   // 写入中的尾行：留给下次
            break
        }
        newOffset = lineOffset + Int64(data.count)
        if let obj = parseJSONLine(Data(data.dropLast())) {
            items.append((lineOffset, obj))
        }
    }
    return (items, newOffset)
}

/// iter_zstd_jsonl：zstd 压缩 JSONL（DSH），通过系统 zstd 二进制解压。
public func iterZstdJSONL(_ path: String) -> [(Int, [String: Any])] {
    let zstd = findTool("zstd")
    let process = Process()
    process.executableURL = URL(fileURLWithPath: zstd)
    process.arguments = ["-d", "-c", path]
    let pipe = Pipe()
    process.standardOutput = pipe
    process.standardError = FileHandle.nullDevice
    do { try process.run() } catch { return [] }
    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    process.waitUntilExit()
    var out: [(Int, [String: Any])] = []
    for (index, line) in splitJSONLines(data).enumerated() {
        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let obj = parseJSONLine(Data(trimmed.utf8)) else { continue }
        out.append((index + 1, obj))
    }
    return out
}

/// _find_tool：PATH → 常见安装目录（.app 图形化启动时 PATH 极简）。
public func findTool(_ name: String) -> String {
    let pathEnv = ProcessInfo.processInfo.environment["PATH"] ?? ""
    for dir in pathEnv.split(separator: ":") {
        let candidate = "\(dir)/\(name)"
        if FileManager.default.isExecutableFile(atPath: candidate) { return candidate }
    }
    let home = NSHomeDirectory()
    for dir in ["/opt/homebrew/bin", "/usr/local/bin",
                "\(home)/.local/bin", "\(home)/.npm-global/bin"] {
        let candidate = "\(dir)/\(name)"
        if FileManager.default.isExecutableFile(atPath: candidate) { return candidate }
    }
    return name
}

// ------------------------------------------------------------ 时间解析 ----

private nonisolated(unsafe) let isoFormatterFractional: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return f
}()

private nonisolated(unsafe) let isoFormatterPlain: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime]
    return f
}()

/// Python datetime.fromisoformat(ts.replace("Z","+00:00")).timestamp()*1000。
public func parseISODateMs(_ ts: String) -> Int64? {
    let normalized = ts.replacingOccurrences(of: "Z", with: "+00:00")
    if let date = isoFormatterFractional.date(from: normalized)
        ?? isoFormatterPlain.date(from: normalized) {
        return Int64(date.timeIntervalSince1970 * 1000)
    }
    return nil
}

// ------------------------------------------------------------ 会话标题 ----

private let contextPrefixes = ["# AGENTS.md", "<INSTRUCTIONS>", "<environment_context>",
                               "<system-reminder>", "Caveat:", "<command-", "<local-command",
                               "<recommended_plugins", "<user_instructions",
                               "## Referenced ChatGPT conversation", "<task-notification",
                               "The following is the Codex agent history"]

private func contentText(_ content: Any?) -> String {
    if let s = content as? String { return s }
    if let list = content as? [Any] {
        return list.compactMap { block -> String? in
            guard let b = block as? [String: Any],
                  let type = b["type"] as? String,
                  type == "text" || type == "input_text" else { return nil }
            return b["text"] as? String ?? ""
        }.joined(separator: " ")
    }
    return ""
}

/// user_text：从 JSONL 行提取首个真实用户消息（跳过 AGENTS.md/环境上下文注入）。
/// 覆盖 claude(type=user)、pi(type=message role=user)、codex(response_item role=user)。
public func userText(_ obj: [String: Any]) -> String {
    var text = ""
    let kind = obj["type"] as? String
    if kind == "user" || kind == "message" {
        if let msg = obj["message"] as? [String: Any], msg["role"] as? String == "user" {
            text = contentText(msg["content"])
        }
    } else if kind == "response_item" {
        if let payload = obj["payload"] as? [String: Any], payload["role"] as? String == "user" {
            text = contentText(payload["content"])
        }
    }
    text = collapseWhitespace(text)
    if text.isEmpty || contextPrefixes.contains(where: { text.hasPrefix($0) }) {
        return ""
    }
    return truncate120(text)
}

// ------------------------------------------------------------ JSON 取值 ----

/// Python `obj.get(k) or 0`：None/缺失/0 → 0；Bool 视为 1/0（Python bool 是 int）。
public func jsonInt(_ value: Any?) -> Int64 {
    switch value {
    case let n as NSNumber:
        return n.int64Value
    case let n as Int64: return n
    case let n as Int: return Int64(n)
    case let n as Double: return Int64(n)
    case let b as Bool: return b ? 1 : 0
    default: return 0
    }
}

/// Python isinstance(value, int) and not bool 且 >= 0（codex _counts 严格校验）。
public func jsonStrictNonNegativeInt(_ value: Any?) -> Int64? {
    guard let n = value as? NSNumber else { return nil }
    if CFGetTypeID(n) == CFBooleanGetTypeID() { return nil }
    if CFNumberIsFloatType(n as CFNumber) { return nil }
    let v = n.int64Value
    return v >= 0 ? v : nil
}

/// Python `a or b`（通用）：a 为 nil/NSNull/空串/0/false 时取 b。
public func jsonOrAny(_ a: Any?, _ b: Any?) -> Any? {
    switch a {
    case nil: return b
    case is NSNull: return b
    case let s as String: return s.isEmpty ? b : a
    case let n as NSNumber:
        if CFGetTypeID(n) == CFBooleanGetTypeID() { return n.boolValue ? a : b }
        return n.doubleValue == 0 ? b : a
    default: return a
    }
}

public func jsonOrAny(_ a: Any?, _ b: Any?, _ c: Any?) -> Any? {
    jsonOrAny(jsonOrAny(a, b), c)
}

public func jsonString(_ value: Any?) -> String? {
    value as? String
}

/// Python `a or b or 0`：第一个真值（非 0 / 非 nil）胜出。
public func jsonOrInt(_ first: Any?, _ second: Any?) -> Int64 {
    let a = jsonInt(first)
    return a != 0 ? a : jsonInt(second)
}

/// Python `a or b`（字符串）：空串 / 缺失回退。
public func jsonOrString(_ first: Any?, _ second: Any?) -> String {
    if let s = first as? String, !s.isEmpty { return s }
    return (second as? String) ?? ""
}


public func jsonOrString(_ first: Any?, _ second: Any?, _ third: Any?) -> String {
    let value = jsonOrString(first, second)
    return value.isEmpty ? (third as? String ?? "") : value
}

/// Stable identifier extraction for JSON values that may encode ids as either
/// strings or JSON numbers.  Numeric sequence ids must not collapse to one
/// empty call id when they are used as activity identities.
public func jsonIdentifier(_ values: Any?...) -> String {
    for value in values {
        if let string = value as? String, !string.isEmpty { return string }
        if let number = value as? NSNumber,
           CFGetTypeID(number) != CFBooleanGetTypeID() {
            return number.stringValue
        }
    }
    return ""
}

File: windows/src/TokenTracker.Core/Scanners/ScannerSupport.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using TokenTracker.Core.Json;
using TokenTracker.Core.Platform;
using TokenTracker.Core.Store;
using ZstdSharp;

namespace TokenTracker.Core.Scanners;

/// <summary>一次扫描在单个工作线程上运行；只有计数离开此范围，绝不含内容。</summary>
public sealed class ScanDiagnostics
{
    public int ParseErrors;
    public int ReadErrors;

    [ThreadStatic] static ScanDiagnostics? _current;

    public static ScanDiagnostics? Current => _current;

    public static void Begin() => _current = new ScanDiagnostics();
}

/// <summary>移植 Scanners/ScannerSupport.swift（_util.py）：JSONL 迭代、字节游标增量读、zstd、ISO 时间、会话标题。</summary>
public static class ScannerSupport
{
    // ------------------------------------------------------------ JSONL ----

    /// <summary>json.loads 一行 → 对象（非对象返回 null 并计入解析错误）。</summary>
    public static Dictionary<string, object?>? ParseJsonLine(ReadOnlySpan<byte> line)
    {
        var obj = PyJson.TryParse(line, out var value) ? value as Dictionary<string, object?> : null;
        if (obj is null && ScanDiagnostics.Current is { } d) d.ParseErrors++;
        return obj;
    }

    static List<(int Line, Dictionary<string, object?> Obj)> ParseLines(byte[] data)
    {
        var output = new List<(int, Dictionary<string, object?>)>();
        var text = Encoding.UTF8.GetString(data); // errors="replace"
        var lineNo = 0;
        foreach (var raw in text.Split('\n'))
        {
            lineNo++;
            var trimmed = raw.Trim();
            if (trimmed.Length == 0) continue;
            if (ParseJsonLine(Encoding.UTF8.GetBytes(trimmed)) is { } obj) output.Add((lineNo, obj));
        }
        return output;
    }

    /// <summary>iter_jsonl：(行号从 1 起, 对象)。解析失败的行跳过，行号仍计入。</summary>
    public static List<(int Line, Dictionary<string, object?> Obj)> IterJsonl(string path)
    {
        var data = SharedFile.ReadAllBytes(path);
        if (data is null)
        {
            if (ScanDiagnostics.Current is { } d) d.ReadErrors++;
            return [];
        }
        return ParseLines(data);
    }

    /// <summary>iter_zstd_jsonl：zstd 压缩 JSONL（DSH），进程内解压，不依赖 zstd.exe。</summary>
    public static List<(int Line, Dictionary<string, object?> Obj)> IterZstdJsonl(string path)
    {
        try
        {
            using var file = SharedFile.OpenRead(path);
            using var decompressor = new DecompressionStream(file);
            using var buffer = new MemoryStream();
            decompressor.CopyTo(buffer);
            return ParseLines(buffer.ToArray());
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException or ZstdException)
        {
            if (ScanDiagnostics.Current is { } d) d.ReadErrors++;
            return [];
        }
    }

    /// <summary>
    /// read_jsonl_delta：从字节偏移（必须是行边界）增量读，只解析新增完整行。
    /// 返回 (items, newOffset)；items 为 (行起始偏移, 对象)。偏移失效（截断/轮转/不在行边界）返回 ([], -1)。
    /// </summary>
    public static (List<(long Offset, Dictionary<string, object?> Obj)> Items, long NewOffset) ReadJsonlDelta(
        string path, long offset)
    {
        var items = new List<(long, Dictionary<string, object?>)>();
        try
        {
            using var stream = SharedFile.OpenRead(path);
            var size = stream.Length;
            if (offset < 0 || offset > size) return (items, -1);
            if (offset > 0)
            {
                stream.Seek(offset - 1, SeekOrigin.Begin);
                if (stream.ReadByte() != '\n') return (items, -1);
            }
            stream.Seek(offset, SeekOrigin.Begin);
            var remaining = new byte[size - offset];
            var read = 0;
            while (read < remaining.Length)
            {
                var n = stream.Read(remaining, read, remaining.Length - read);
                if (n == 0) break;
                read += n;
            }
            var newOffset = offset;
            var start = 0;
            while (start < read)
            {
                var nl = Array.IndexOf(remaining, (byte)'\n', start, read - start);
                if (nl < 0) break; // 写入中的尾行：留给下次
                var lineOffset = offset + start;
                newOffset = offset + nl + 1;
                if (ParseJsonLine(remaining.AsSpan(start, nl - start)) is { } obj) items.Add((lineOffset, obj));
                start = nl + 1;
            }
            return (items, newOffset);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            return ([], -1);
        }
    }

    // ------------------------------------------------------------ 时间解析 ----

    static readonly Regex IsoRegex = new(
        @"^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2})(?::(\d{2})(?:[.,](\d+))?)?(Z|[+-]\d{2}(?::?\d{2})?)$",
        RegexOptions.CultureInvariant);

    /// <summary>
    /// Python datetime.fromisoformat(ts.replace("Z","+00:00")).timestamp()*1000（截断）。
    /// 与 Swift 一样要求带时区；无时区返回 null。
    /// </summary>
    public static long? ParseIsoDateMs(string ts)
    {
        var m = IsoRegex.Match(ts.Trim());
        if (!m.Success) return null;
        try
        {
            int G(int i) => m.Groups[i].Success ? int.Parse(m.Groups[i].Value, CultureInfo.InvariantCulture) : 0;
            var zone = m.Groups[8].Value;
            var offset = TimeSpan.Zero;
            if (zone != "Z")
            {
                var sign = zone[0] == '-' ? -1 : 1;
                var digits = zone[1..].Replace(":", "");
                var hours = int.Parse(digits[..2], CultureInfo.InvariantCulture);
                var minutes = digits.Length >= 4 ? int.Parse(digits[2..4], CultureInfo.InvariantCulture) : 0;
                offset = TimeSpan.FromMinutes(sign * (hours * 60 + minutes));
            }
            var dto = new DateTimeOffset(G(1), G(2), G(3), G(4), G(5), G(6), offset);
            var fraction = m.Groups[7].Success ? m.Groups[7].Value : "";
            var micros = fraction.Length == 0 ? 0 : long.Parse(fraction.PadRight(6, '0')[..6], CultureInfo.InvariantCulture);
            var totalMicros = dto.ToUnixTimeSeconds() * 1_000_000 + micros;
            return (long)(totalMicros / 1e6 * 1000);
        }
        catch (ArgumentOutOfRangeException)
        {
            return null;
        }
    }

    /// <summary>数字（秒或毫秒）或 ISO 字符串 → 毫秒；无法解析返回 0。</summary>
    public static long ParseTs(object? raw)
    {
        if (PyJson.AsNumber(raw) is { } d) return (long)(d < 1e12 ? d * 1000 : d);
        if (raw is string s) return ParseIsoDateMs(s) ?? 0;
        return 0;
    }

    // ------------------------------------------------------------ 会话标题 ----

    static readonly string[] ContextPrefixes =
    [
        "# AGENTS.md", "<INSTRUCTIONS>", "<environment_context>", "<system-reminder>", "Caveat:", "<command-",
        "<local-command", "<recommended_plugins", "<user_instructions", "## Referenced ChatGPT conversation",
        "<task-notification", "The following is the Codex agent history",
    ];

    static string ContentText(object? content) => content switch
    {
        string s => s,
        List<object?> list => string.Join(" ", list.OfType<Dictionary<string, object?>>()
            .Where(b => b.Get("type") is "text" or "input_text")
            .Select(b => b.TryGetValue("text", out var t) ? PyJson.PyStr(t) : "")),
        _ => "",
    };

    /// <summary>user_text：首个真实用户消息（跳过 AGENTS.md / 环境上下文注入）。</summary>
    public static string UserText(Dictionary<string, object?> obj)
    {
        var text = "";
        var kind = obj.Get("type") as string;
        if (kind is "user" or "message")
        {
            if (obj.Get("message") is Dictionary<string, object?> msg && msg.Get("role") as string == "user")
                text = ContentText(msg.Get("content"));
        }
        else if (kind == "response_item")
        {
            if (obj.Get("payload") is Dictionary<string, object?> payload && payload.Get("role") as string == "user")
                text = ContentText(payload.Get("content"));
        }
        text = TextUtil.CollapseWhitespace(text);
        if (text.Length == 0 || ContextPrefixes.Any(p => text.StartsWith(p, StringComparison.Ordinal))) return "";
        return TextUtil.Truncate120(text);
    }
}
